File: js-for-python-programmers/classes.js

var inherits = require('util').inherits;

var Foo = function (bar) {
    console.log('Creating a new Foo');
    this.bar = bar;
};

Foo.prototype.zoop = function () {
    console.log('Foo.zoop invoked: bar = ' + this.bar);
};

Foo.prototype.zap = function () {
    console.log('Foo.zap invoked: bar = ' + this.bar);
};

var Quux = function (bar) {
    console.log('Creating a new Quux');
    Foo.apply(this, [bar]);
};
inherits(Quux, Foo);

Quux.prototype.zoop = function () {
    console.log('Quux.zoop invoked.  Invoking Foo.zoop...');
    Foo.prototype.zoop.apply(this);
};

if (module.parent === null) {
    console.log('Starting a javascript program.');
    var foo = new Foo('baz');
    foo.zoop();

    var quux = new Quux('quuxBaz');
    quux.zoop();
    quux.zap();
}

File: js-for-python-programmers/classes.py

class Foo(object):

    def __init__(self, bar):
        print 'Creating a new Foo'
        self.bar = bar

    def zoop(self):
        print 'Foo.zoop invoked: bar = ' + self.bar 

    def zap(self):
        print 'Foo.zap invoked: bar = ' + self.bar 


class Quux(Foo):

    def __init__(self, bar):
        print 'Creating a new Quux'
        Foo.__init__(self, bar)
        
    def zoop(self):
        print 'Quux.zoop invoked.  Invoking Foo.zoop...'
        Foo.zoop(self)

if __name__ == '__main__':
    print 'Starting a python program.'
    foo = Foo('baz')
    foo.zoop()
    foo.zap()
    quux = Quux('quuxBaz')
    quux.zoop()
    quux.zap()

        
